common/cpp/src/google_smart_card_common/value_emscripten_val_conversion_unittest.cc

```diff
@@ -0,0 +1,205 @@
+  EXPECT_TRUE(converted.isNull());
+}
+
+TEST(ValueEmscriptenValConversion, BooleanValue) {
+  {
+    constexpr bool kBoolean = false;
+    const emscripten::val converted =
+        ConvertValueToEmscriptenVal(Value(kBoolean));
+    ASSERT_EQ(converted.typeOf().as<std::string>(), "boolean");
+    EXPECT_EQ(converted.as<bool>(), kBoolean);
+  }
+
+  {
+    constexpr bool kBoolean = true;
+    const emscripten::val converted =
+        ConvertValueToEmscriptenVal(Value(kBoolean));
+    ASSERT_EQ(converted.typeOf().as<std::string>(), "boolean");
+    EXPECT_EQ(converted.as<bool>(), kBoolean);
+  }
+}
+
+TEST(ValueEmscriptenValConversion, IntegerValue) {
+  constexpr int kNumber = 123;
+  const emscripten::val converted = ConvertValueToEmscriptenVal(Value(kNumber));
+  ASSERT_TRUE(converted.isNumber());
+  EXPECT_EQ(converted.as<int>(), kNumber);
+}
+
+TEST(ValueEmscriptenValConversion, IntegerNon32BitValue) {
+  constexpr int64_t k40Bit = 1LL << 40;
+  const emscripten::val converted = ConvertValueToEmscriptenVal(Value(k40Bit));
+  ASSERT_TRUE(converted.isNumber());
+  // `emscripten::val` doesn't provide a direct way to transform into a
+  // non-32-bit integer, so compare string representations.
+  EXPECT_EQ(converted.call<std::string>("toString"), std::to_string(k40Bit));
+}
+
+TEST(ValueEmscriptenValConversion, Integer64BitMaxValue) {
+  constexpr int64_t k64BitMax = std::numeric_limits<int64_t>::max();
+  const emscripten::val converted =
+      ConvertValueToEmscriptenVal(Value(k64BitMax));
+  ASSERT_TRUE(converted.isNumber());
+  EXPECT_DOUBLE_EQ(converted.as<double>(), static_cast<double>(k64BitMax));
+}
+
+TEST(ValueEmscriptenValConversion, Integer64BitMinValue) {
+  constexpr int64_t k64BitMin = std::numeric_limits<int64_t>::min();
+  const emscripten::val converted =
+      ConvertValueToEmscriptenVal(Value(k64BitMin));
+  ASSERT_TRUE(converted.isNumber());
+  EXPECT_DOUBLE_EQ(converted.as<double>(), static_cast<double>(k64BitMin));
+}
+
+TEST(ValueEmscriptenValConversion, FloatValue) {
+  constexpr double kFloat = 123.456;
+  const emscripten::val converted = ConvertValueToEmscriptenVal(Value(kFloat));
+  ASSERT_TRUE(converted.isNumber());
+  EXPECT_DOUBLE_EQ(converted.as<double>(), static_cast<double>(kFloat));
+}
+
+TEST(ValueEmscriptenValConversion, StringValue) {
+  {
+    const emscripten::val converted =
+        ConvertValueToEmscriptenVal(Value(Value::Type::kString));
+    ASSERT_TRUE(converted.isString());
+    EXPECT_EQ(converted.as<std::string>(), "");
+  }
+
+  {
+    const char kFoo[] = "foo";
+    const emscripten::val converted = ConvertValueToEmscriptenVal(Value(kFoo));
+    ASSERT_TRUE(converted.isString());
+    EXPECT_EQ(converted.as<std::string>(), kFoo);
+  }
+}
+
+TEST(ValueEmscriptenValConversion, BinaryValue) {
+  {
+    const emscripten::val converted =
+        ConvertValueToEmscriptenVal(Value(Value::Type::kBinary));
+    ASSERT_TRUE(converted.instanceof (emscripten::val::global("ArrayBuffer")));
```

Comment:
Hmm, this looks like a clang-format bug. This `instanceof()` is a regular C++ class method, there shouldn't be a space after it. I guess we can keep it as it is. The alternative is to remove the space and surround this code with `// clang-format off`.

Reply:
I am leaning towards disabling clang-format in this line. It feels like a special operator while in fact it is a class method. Or maybe clang has an rational explanation for this? (I couldn't find any)